runtime/onert/core/include/ir/Model.h

```diff
@@ -177,23 +177,30 @@ class Model
   std::shared_ptr<backend::custom::IKernelBuilder> _kernel_builder;
 
 public:
-  void add_metadata(const std::string &name, std::shared_ptr<const ir::Data> data)
+  void add_metadata(const std::string &name, std::unique_ptr<const ir::Data> data)
   {
-    _metadatas.emplace(name, data);
+    _metadatas.emplace(name, std::move(data));
   }
 
-  bool is_metadata_exist(const std::string &name) const
+  bool exists_metadata(const std::string &name) const
   {
     return _metadatas.find(name) != _metadatas.end();
   }
 
-  std::shared_ptr<const ir::Data> get_metadata(const std::string name) const
+  std::unique_ptr<const ir::Data> get_metadata(const std::string name)
   {
-    return _metadatas.at(name);
+    auto m = _metadatas.find(name);
+
+    if (m == _metadatas.end())
+      throw std::out_of_range{"no meatdata named " + name};
+
+    auto data = std::move(m->second);
+    _metadatas.erase(m);
```

Comment:
@zetwhite It doesn't look intuitive to erase the metadata after `get_metadata`. Do we need to return `unique_ptr`?

Reply:
Umm, I've also thought about returning "const raw ptr" instead of `unique_ptr`.

I think that returning `unique_ptr` is much safer.
I'm a little afraid that under situation happened with "const raw ptr". 

```cpp

 // load model from file
std::unique_ptr<ir::Model> model = laodModel();
 
// get buffer
const ir::Data* metadata_buffer = model->get_metadata("CIRCLE_TRAINING"); 

// delete a model
// Since the model holds unique_ptr of metadata, all metadata are also deleted.
model.reset(); 

// access already freed memory space
metadata_buffer->base
```

Reply:
Since `ir::Model` doesn't know how methods can be used, I thought it is better to use `unique_ptr` for safety. 
Do you have better idea about this, plz let me know! 

+) If erasing metadata after`get_metadata` is un-intutive, we could find a bettter naming instead of `get_metadata`- :thinking: :thinking:  maybe `extract_metadata`?

Reply:
`extract_metadata` looks better to me.

Reply:
I don't object to use `unique_ptr`. `unique_ptr` is good thing to prevent memory leak. I just don't get the user surprised the metadata is removed after `get_metadata`. If you choose to remove metadata (though I don't like it), it would be better to rename which says clearly the metadata will not be available once you get it. At least for me, `extract` is not good one. I tried to find more appropriate word, but I don't have idea. 

Once again, I don't oppose this PR if it gets two approvals. It will be merged.

Reply:
> How about introducing a class for metadata into onert? If so, you can prevent the problem you worried about by adjusting the introduced class (i.g. introducing getting const reference of metadata class into ir::Model and deleting copy constructor of metadata class). Furthermore, you can take advantage of easily adding functions related to metadata.

 @ragmani Thanks for thinking together. I failed to fully understand your suggestion. 
If we return a 'const reference of metadata' from `get_metadata`, Does it extend the lifetime of metadata, even if the model instance is deleted? 

> Furthermore, you can take advantage of easily adding functions related to metadata. 

(not important, but for sharing opinion) 
Umm.. I think there are not many functions needed for metadata. Since metadata in a model is just a buffer(not parsed yet), there aren't many common things between metadata. 
So, This advantage is not very attractive to me.

Reply:
For now, I'd like to rename `get_metadata` to `extract_metadata` and add some comments for `ir::Model` user.

If I could get 2 approvals, let it proceed. 
Or Let me try to find another solution. 

(note) I get the word `extract` from c++17 std.
>  unlinks the node that contains the element pointed to by position and returns a node handle that owns it. 
(https://en.cppreference.com/w/cpp/container/map/extract)

Reply:
Wow, I didn't know `map.extract()`. Then, we may replace the body of `extract_metadata`:
```cc
    auto m = _metadatas.find(name);

    if (m == _metadatas.end())
      throw std::out_of_range{"no meatdata named " + name};

    auto data = std::move(m->second);
    _metadatas.erase(m);
    return data;
```
with:
```cc
   // After c++17
   return std::move(_metadatas.extract(name).mapped());
```

Reply:
> If we return a 'const reference of metadata' from get_metadata, Does it extend the lifetime of metadata, even if the model instance is deleted?

No, it does not extend lifetime of metadata. If the reference exists and the model instance is deleted, the reference is a dangling reference.

> (not important, but for sharing opinion)
Umm.. I think there are not many functions needed for metadata. Since metadata in a model is just a buffer(not parsed yet), there aren't many common things between metadata.
So, This advantage is not very attractive to me.

I see, it was just a suggestion.